ckks/encryptor.go

```diff
@@ -3,394 +3,54 @@ package ckks
-
-		// ct1 = u*pk1 + e1
-		encryptor.gaussianSampler.ReadLvl(lvl, poolQ0)
-		ringQ.NTTLvl(lvl, poolQ0, poolQ0)
-		ringQ.AddLvl(lvl, ciphertext.Value[1], poolQ0, ciphertext.Value[1])
-
-		if !plaintext.IsNTT {
-
-			// ct0 = u*pk0 + e0
-			encryptor.gaussianSampler.ReadLvl(lvl, poolQ0)
-			// ct0 = (u*pk0 + e0)/P + m
-			ringQ.AddLvl(lvl, poolQ0, plaintext.Value, poolQ0)
-			ringQ.NTTLvl(lvl, poolQ0, poolQ0)
-			ringQ.AddLvl(lvl, ciphertext.Value[0], poolQ0, ciphertext.Value[0])
-
-		} else {
-			// ct0 = u*pk0 + e0
-			encryptor.gaussianSampler.ReadLvl(lvl, poolQ0)
-			ringQ.NTTLvl(lvl, poolQ0, poolQ0)
-			ringQ.AddLvl(lvl, ciphertext.Value[0], poolQ0, ciphertext.Value[0])
-			ringQ.AddLvl(lvl, ciphertext.Value[0], plaintext.Value, ciphertext.Value[0])
-		}
-
-	} else {
-
-		ringP := encryptor.ringP
-
-		encryptor.ternarySampler.ReadLvl(lvl, poolQ2)
-
-		extendBasisSmallNormAndCenter(ringQ, ringP, poolQ2, poolP2)
-
-		// (#Q + #P) NTT
-		ringQ.NTTLvl(lvl, poolQ2, poolQ2)
-		ringP.NTT(poolP2, poolP2)
-
-		ringQ.MFormLvl(lvl, poolQ2, poolQ2)
-		ringP.MForm(poolP2, poolP2)
-
-		pk0P := new(ring.Poly)
-		pk1P := new(ring.Poly)
-		pk0P.Coeffs = encryptor.pk.Value[0].Coeffs[len(ringQ.Modulus):]
-		pk1P.Coeffs = encryptor.pk.Value[1].Coeffs[len(ringQ.Modulus):]
-
-		// ct0 = u*pk0
-		// ct1 = u*pk1
-		ringQ.MulCoeffsMontgomeryLvl(lvl, poolQ2, encryptor.pk.Value[0], poolQ0)
-		ringQ.MulCoeffsMontgomeryLvl(lvl, poolQ2, encryptor.pk.Value[1], poolQ1)
-		ringP.MulCoeffsMontgomery(poolP2, pk0P, poolP0)
-		ringP.MulCoeffsMontgomery(poolP2, pk1P, poolP1)
-
-		// 2*(#Q + #P) NTT
-		ringQ.InvNTTLvl(lvl, poolQ0, poolQ0)
-		ringQ.InvNTTLvl(lvl, poolQ1, poolQ1)
-		ringP.InvNTT(poolP0, poolP0)
-		ringP.InvNTT(poolP1, poolP1)
-
-		// ct0 = u*pk0 + e0
-		encryptor.gaussianSampler.ReadLvl(lvl, poolQ2)
-		extendBasisSmallNormAndCenter(ringQ, ringP, poolQ2, poolP2)
-		ringQ.AddLvl(lvl, poolQ0, poolQ2, poolQ0)
-		ringP.Add(poolP0, poolP2, poolP0)
-
-		// ct1 = u*pk1 + e1
-		encryptor.gaussianSampler.ReadLvl(lvl, poolQ2)
-		extendBasisSmallNormAndCenter(ringQ, ringP, poolQ2, poolP2)
-		ringQ.AddLvl(lvl, poolQ1, poolQ2, poolQ1)
-		ringP.Add(poolP1, poolP2, poolP1)
-
-		// ct0 = (u*pk0 + e0)/P
-		encryptor.baseconverter.ModDownSplitPQ(lvl, poolQ0, poolP0, ciphertext.Value[0])
-
-		// ct1 = (u*pk1 + e1)/P
-		encryptor.baseconverter.ModDownSplitPQ(lvl, poolQ1, poolP1, ciphertext.Value[1])
-
-		if !plaintext.IsNTT {
-			ringQ.AddLvl(lvl, ciphertext.Value[0], plaintext.Value, ciphertext.Value[0])
-		}
-
-		// 2*#Q NTT
-		ringQ.NTTLvl(lvl, ciphertext.Value[0], ciphertext.Value[0])
-		ringQ.NTTLvl(lvl, ciphertext.Value[1], ciphertext.Value[1])
-
-		if plaintext.IsNTT {
-			// ct0 = (u*pk0 + e0)/P + m
-			ringQ.AddLvl(lvl, ciphertext.Value[0], plaintext.Value, ciphertext.Value[0])
-		}
+// NewEncryptor instanciate a new wrapped rlwe.Encryptor for CKKS ciphertexts and plaintexts.
+func NewEncryptor(params Parameters, key interface{}) *Encryptor {
+	switch key := key.(type) {
```

Comment:
Can this switch be done at the rlwe layer ?

Reply:
fixed